modules/nf-core/limma/differential/templates/limma_de_rnaseq.R

```diff
@@ -0,0 +1,411 @@
+
+print(opt)
+
+# Set correct types for options
+for (name in names(opt)) {
+    if (!is.null(opt[[name]])) {
+        class(opt[[name]]) <- opt_types[[name]]
+    }
+}
+
+## CHECKS
+# Check if required parameters have been provided
+required_opts <- c('contrast_variable', 'reference_level', 'target_level', 'output_prefix')
+missing <- required_opts[unlist(lapply(opt[required_opts], is.null)) | ! required_opts %in% names(opt)]
+
+if (length(missing) > 0){
+    stop(paste("Missing required options:", paste(missing, collapse=', ')))
+}
+
+# Check file inputs are valid
+for (file_input in c('count_file', 'sample_file')){
+    if (is.null(opt[[file_input]])) {
+        stop(paste("Please provide", file_input), call. = FALSE)
+    }
+
+    if (! file.exists(opt[[file_input]])){
+        stop(paste0('Value of ', file_input, ': ', opt[[file_input]], ' is not a valid file'))
+    }
+}
+
+# Convert things to vectors where needed
+vector_opt <- c('stdev.coef.lim', 'winsor.tail.p')
+opt[vector_opt] = lapply(strsplit(unlist(opt[vector_opt]), ','), as.numeric)
+
+# READ IN COUNTS FILE AND SAMPLE METADATA
+count.table <-
+    read_delim_flexible(
+        file = opt\$count_file,
+        header = TRUE,
+        row.names = 1,
+        check.names = FALSE
+    )
+sample.sheet <- read_delim_flexible(file = opt\$sample_file)
+
+# Correct sample.sheet colnames
+colnames(sample.sheet) <- make.names(colnames(sample.sheet))
+
+# Deal with spaces that may be in sample column
+opt\$sample_id_col <- make.names(opt\$sample_id_col)
+
+if (! opt\$sample_id_col %in% colnames(sample.sheet)){
+    stop(paste0("Specified sample ID column '", opt\$sample_id_col, "' is not in the sample sheet"))
+}
+
+# Sample sheet can have duplicate rows for multiple sequencing runs, so uniqify
+# before assigning row names
+sample.sheet <- sample.sheet[! duplicated(sample.sheet[[opt\$sample_id_col]]), ]
+rownames(sample.sheet) <- sample.sheet[[opt\$sample_id_col]]
+
+# Check that all samples specified in the input sheet are present in the
+# counts table. Assuming they are, subset and sort the count table to
+# match the sample sheet
+missing_samples <-
+    sample.sheet[!rownames(sample.sheet) %in% colnames(count.table), opt\$sample_id_col]
+
+if (length(missing_samples) > 0) {
+    stop(paste(
+        length(missing_samples),
+        'specified samples missing from count table:',
+        paste(missing_samples, collapse = ',')
+    ))
+} else{
+    # Save any non-count data, will gene metadata etc we might need later
+    noncount.table <-
+        count.table[, !colnames(count.table) %in% rownames(sample.sheet), drop = FALSE]
+    count.table <- count.table[, rownames(sample.sheet)]
+}
+
+########## START ANALYSIS
+DGE <- DGEList(count.table)
+dim(DGE)
+
+## Filter genes
+keep <- which(apply(DGE\$counts, 1, max) >= opt\$geneFilter)
+filtered_counts <- DGE\$counts[keep, ]
+DGE\$counts <- filtered_counts
+dim(DGE)
+
+# Calculate normalization factors
+DGE <- calcNormFactors(DGE)
```

Comment:
Please correct me if I'm wrong, but this and the DGEList are the only things that need edgeR here, right? voom() will do the normalisation things, and will just take a matrix. Then we don't need the edgeR dependency.

Reply:
We use edgeR dependency to read raw matrix count to create DGEList object which is one of required ways to provide expression data to voom: "counts - a numeric matrix containing raw counts, or an ExpressionSet containing raw counts, or a DGEList object." : https://www.rdocumentation.org/packages/limma/versions/3.28.14/topics/voom
 
 So we either need BioBase or edgeR.

Reply:
Yep, as I say, we don't need edgeR, and that will simplify the software dependencies